cumulus/primitives/utility/src/lib.rs

```diff
@@ -286,23 +291,238 @@ impl<
+		AccountId,
+	)>,
+}
+
+impl<
+		Target: Get<Fungibles::AssetId>,
+		SwapCredit: SwapCreditT<
+			AccountId,
+			Balance = Fungibles::Balance,
+			AssetKind = Fungibles::AssetId,
+			Credit = fungibles::Credit<AccountId, Fungibles>,
+		>,
+		WeightToFee: WeightToFeeT<Balance = Fungibles::Balance>,
+		Fungibles: fungibles::Balanced<AccountId>,
+		FungiblesAssetMatcher: MatchesFungibles<Fungibles::AssetId, Fungibles::Balance>,
+		OnUnbalanced: OnUnbalancedT<fungibles::Credit<AccountId, Fungibles>>,
+		AccountId,
+	> WeightTrader
+	for SwapFirstAssetTrader<
+		Target,
+		SwapCredit,
+		WeightToFee,
+		Fungibles,
+		FungiblesAssetMatcher,
+		OnUnbalanced,
+		AccountId,
+	> where
+	Fungibles::Balance: Into<u128>,
+{
+	fn new() -> Self {
+		Self {
+			total_fee: fungibles::Credit::<AccountId, Fungibles>::zero(Target::get()),
+			last_fee_asset: None,
+			_phantom_data: PhantomData,
+		}
+	}
+
+	fn buy_weight(
+		&mut self,
+		weight: Weight,
+		mut payment: xcm_executor::Assets,
+		_context: &XcmContext,
+	) -> Result<xcm_executor::Assets, XcmError> {
+		log::trace!(
+			target: "xcm::weight",
+			"SwapFirstAssetTrader::buy_weight weight: {:?}, payment: {:?}",
+			weight,
+			payment,
+		);
+		let first_asset: MultiAsset =
+			payment.fungible.pop_first().ok_or(XcmError::AssetNotFound)?.into();
+		let (fungibles_asset, balance) = FungiblesAssetMatcher::matches_fungibles(&first_asset)
+			.map_err(|_| XcmError::AssetNotFound)?;
+
+		let swap_asset = fungibles_asset.clone().into();
+		if Target::get().eq(&swap_asset) {
+			// current trader is not applicable.
+			return Err(XcmError::FeesNotMet)
+		}
+
+		let credit_in = Fungibles::issue(fungibles_asset, balance);
+		let fee = WeightToFee::weight_to_fee(&weight);
+
+		// swap the user's asset for the `Target` asset.
+		let (credit_out, credit_change) = SwapCredit::swap_tokens_for_exact_tokens(
+			vec![swap_asset, Target::get()],
+			credit_in,
+			fee,
+		)
+		.map_err(|(credit_in, _)| {
+			drop(credit_in);
+			XcmError::FeesNotMet
+		})?;
+
+		match self.total_fee.subsume(credit_out) {
+			Err(credit_out) => {
+				// error may occur if `total_fee.asset` differs from `credit_out.asset`, which does
+				// not apply in this context.
+				defensive!(
+					"`total_fee.asset` must be equal to `credit_out.asset`",
+					(self.total_fee.asset(), credit_out.asset())
+				);
+				return Ok(payment)
+			},
+			_ => (),
+		};
+		self.last_fee_asset = Some(first_asset.id);
+
+		payment.fungible.insert(first_asset.id, credit_change.peek().into());
+		drop(credit_change);
+		Ok(payment)
+	}
+
+	fn refund_weight(&mut self, weight: Weight, _context: &XcmContext) -> Option<MultiAsset> {
+		log::trace!(
+			target: "xcm::weight",
+			"SwapFirstAssetTrader::refund_weight weight: {:?}, self.total_fee: {:?}",
+			weight,
+			self.total_fee,
+		);
+		if self.total_fee.peek().is_zero() {
+			// noting yet paid to refund.
+			return None
+		}
+		let mut refund_asset = if let Some(asset) = &self.last_fee_asset {
```

Comment:
This `last_fee_asset` seems a bit odd to me, since it would not work multiple times. Eg:  
```js
buy()
buy()
refund()
refund() // uses a different asset here
```
is this intended?

Reply:
yes, I chose a simper solution here. if you paid with asset `a` first, and `b` later. we do the refund with the last asset you used. this should cover most of the cases.
also `refund_weight`, returns only a single `MultiAsset` is succeeded.

Reply:
Okay I see. @franciscoaguirre do you know if this is according to the XCM specs?

Reply:
Nothing in the XCM spec says it can't be like that. Since it only returns one asset, we're stuck with doing something like this. It would have to be well documented, that's all.